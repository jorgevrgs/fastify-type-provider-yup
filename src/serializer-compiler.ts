import type { FastifySerializerCompiler } from "fastify/types/schema";
import type { AnySchema } from "yup";
import { defaultYupValidatorCompilerOptions } from "./constants";
import { ResponseValidationError } from "./errors";
import type { YupValidatorCompilerOptions } from "./types";
import { resolveSchema, safeParse } from "./utils";

export const createSerializerCompiler = (
  options: YupValidatorCompilerOptions,
) => {
  const serializerCompiler: FastifySerializerCompiler<
    AnySchema | { properties: AnySchema }
  > = ({ schema: maybeSchema }) => {
    // biome-ignore lint/suspicious/noExplicitAny: required for type casting
    return (data: any) => {
      const schema = resolveSchema(maybeSchema);

      const result = safeParse(schema, data, options);

      if (result.success === true) {
        return JSON.stringify(result.data);
      }

      throw new ResponseValidationError(result);
    };
  };

  return serializerCompiler;
};

export const serializerCompiler = createSerializerCompiler(
  defaultYupValidatorCompilerOptions,
);
