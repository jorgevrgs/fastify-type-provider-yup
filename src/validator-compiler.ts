import type { FastifySchemaCompiler } from 'fastify';
import type { AnySchema } from 'yup';
import { defaultYupValidatorCompilerOptions } from './constants';
import type { YupValidatorCompilerOptions } from './types';

export const createValidatorCompiler = (options: YupValidatorCompilerOptions) => {
  const validatorCompiler: FastifySchemaCompiler<AnySchema> = ({ schema }) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return async (data: any) => {
      try {
        const value = schema.validateSync(data, options);

        return {
          value,
        };
      } catch (error) {
        return {
          error,
        };
      }
    };
  };

  return validatorCompiler;
};

export const validatorCompiler = createValidatorCompiler(defaultYupValidatorCompilerOptions);
