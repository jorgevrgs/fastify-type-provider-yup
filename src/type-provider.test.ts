import type { FastifyInstance } from 'fastify';
import Fastify from 'fastify';
import { afterAll, beforeAll, beforeEach, describe, expect, it } from 'vitest';
import * as yup from 'yup';
import type { YupTypeProvider } from './type-provider';

import { serializerCompiler } from './serializer-compiler';
import { validatorCompiler } from './validator-compiler';

describe('response schema', () => {
  let app: FastifyInstance;

  beforeAll(async () => {
    const REQUEST_SCHEMA = yup.object({
      name: yup.string().required(),
    });

    app = Fastify();
    app.setValidatorCompiler(validatorCompiler);
    app.setSerializerCompiler(serializerCompiler);

    app.after(() => {
      app
        .withTypeProvider<YupTypeProvider>()
        .route({
          method: 'GET',
          url: '/with-schema',
          schema: {
            querystring: REQUEST_SCHEMA,
          },
          handler: (req, res) => {
            res.send({
              name: req.query.name,
            });
          },
        })
        .route({
          method: 'GET',
          url: '/without-schema',
          schema: undefined,
          handler: (req, res) => {
            res.send({
              status: 'ok',
            });
          },
        });
    });

    await app.ready();
  });

  afterAll(async () => {
    await app.close();
  });

  it('accepts correct request', async () => {
    const response = await app.inject({
      method: 'GET',
      url: '/with-schema',
      query: {
        name: 'test',
      },
    });

    expect(response.statusCode).toBe(200);
    expect(response.json()).toEqual({
      name: 'test',
    });
  });

  it('accepts request on route without schema', async () => {
    const response = await app.inject({
      method: 'GET',
      url: '/without-schema',
    });

    expect(response.statusCode).toBe(200);
    expect(response.json()).toStrictEqual({
      status: 'ok',
    });
  });

  it('returns 400 on validation error', async () => {
    const response = await app.inject({
      method: 'GET',
      url: '/with-schema',
    });

    expect(response.statusCode).toBe(400);
    expect(response.body).toMatchSnapshot();
  });
});

describe('response schema', () => {
  describe('does not fail on empty response schema (204)', () => {
    let app: FastifyInstance;

    beforeAll(async () => {
      app = Fastify();
      app.setValidatorCompiler(validatorCompiler);
      app.setSerializerCompiler(serializerCompiler);

      app.after(() => {
        app
          .withTypeProvider<YupTypeProvider>()
          .route({
            method: 'GET',
            url: '/',
            schema: {
              response: {
                204: yup.mixed().oneOf([undefined], 'test'),
              },
            },
            handler: (req, res) => {
              res.status(204).send();
            },
          })
          .route({
            method: 'GET',
            url: '/incorrect',
            schema: {
              response: {
                204: yup.mixed().oneOf([undefined], 'test'),
              },
            },
            handler: (req, res) => {
              res.status(204).send({ id: 1 });
            },
          });
      });
      await app.ready();
    });

    afterAll(async () => {
      await app.close();
    });

    it('returns 204', async () => {
      const response = await app.inject().get('/');

      expect(response.statusCode).toBe(204);
      expect(response.body).toEqual('');
    });

    it('throws on non-empty', async () => {
      const response = await app.inject().get('/incorrect');

      expect(response.statusCode).toBe(500);
      expect(response.json()).toEqual({
        error: 'Internal Server Error',
        message: "Response doesn't match the schema",
        statusCode: 500,
      });
    });
  });

  describe('correctly processes response schema (string)', () => {
    let app: FastifyInstance;
    beforeAll(async () => {
      const REPLY_SCHEMA = yup.string();

      app = Fastify();
      app.setValidatorCompiler(validatorCompiler);
      app.setSerializerCompiler(serializerCompiler);

      app.after(() => {
        app.withTypeProvider<YupTypeProvider>().route({
          method: 'GET',
          url: '/',
          schema: {
            response: {
              200: REPLY_SCHEMA,
            },
          },
          handler: (req, res) => {
            res.send('test');
          },
        });

        app.withTypeProvider<YupTypeProvider>().route({
          method: 'GET',
          url: '/incorrect',
          schema: {
            response: {
              200: REPLY_SCHEMA,
            },
          },
          handler: (req, res) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            res.send({ name: 'test' } as any);
          },
        });
      });

      await app.ready();
    });

    afterAll(async () => {
      await app.close();
    });

    it('returns 200 on correct response', async () => {
      const response = await app.inject().get('/');

      expect(response.statusCode).toBe(200);
      expect(response.body).toEqual('test');
    });

    it('returns 500 on incorrect response', async () => {
      const response = await app.inject().get('/incorrect');

      expect(response.statusCode).toBe(500);
      expect(response.body).toMatchSnapshot();
    });
  });

  describe('correctly processes response schema (object)', () => {
    let app: FastifyInstance;
    beforeEach(async () => {
      const REPLY_SCHEMA = yup.object({
        name: yup.string(),
      });

      app = Fastify();
      app.setValidatorCompiler(validatorCompiler);
      app.setSerializerCompiler(serializerCompiler);

      app.after(() => {
        app.withTypeProvider<YupTypeProvider>().route({
          method: 'GET',
          url: '/',
          schema: {
            response: {
              200: REPLY_SCHEMA,
            },
          },
          handler: (req, res) => {
            res.send({
              name: 'test',
            });
          },
        });

        app.withTypeProvider<YupTypeProvider>().route({
          method: 'GET',
          url: '/incorrect',
          schema: {
            response: {
              200: REPLY_SCHEMA,
            },
          },
          handler: (req, res) => {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            res.send('test' as any);
          },
        });
      });

      await app.ready();
    });
    afterAll(async () => {
      await app.close();
    });

    it('returns 200 for correct response', async () => {
      const response = await app.inject().get('/');

      expect(response.statusCode).toBe(200);
      expect(response.json()).toEqual({
        name: 'test',
      });
    });

    // FixMe https://github.com/turkerdev/fastify-type-provider-zod/issues/16
    it.skip('returns 500 for incorrect response', async () => {
      const response = await app.inject().get('/incorrect');

      expect(response.statusCode).toBe(500);
      expect(response.json()).toMatchSnapshot();
    });
  });
});
