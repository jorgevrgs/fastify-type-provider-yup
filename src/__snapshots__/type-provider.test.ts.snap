// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`response schema > correctly processes response schema (string) > returns 500 on incorrect response 1`] = `"{\\"statusCode\\":500,\\"error\\":\\"Internal Server Error\\",\\"message\\":\\"Response doesn't match the schema\\"}"`;

exports[`response schema > returns 400 on validation error 1`] = `"{\\"statusCode\\":400,\\"code\\":\\"FST_ERR_VALIDATION\\",\\"error\\":\\"Bad Request\\",\\"message\\":\\"name is a required field\\"}"`;
